datanode/dehistory/protocol_upgrade_handler.go

```diff
@@ -0,0 +1,81 @@
+package dehistory
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"code.vegaprotocol.io/vega/datanode/dehistory/snapshot"
+	"code.vegaprotocol.io/vega/datanode/service"
+	"code.vegaprotocol.io/vega/logging"
+)
+
+type ProtocolUpgradeHandler struct {
+	log                          *logging.Logger
+	cfg                          Config
+	createSnapshotAsynchronously func(ctx context.Context, chainID string, toHeight int64) (snapshot.MetaData, error)
+	protocolUpgradeService       *service.ProtocolUpgrade
+	snapshotPublished            func(meta snapshot.MetaData) (bool, error)
+}
+
+func NewProtocolUpgradeHandler(log *logging.Logger, cfg Config, protocolUpgradeService *service.ProtocolUpgrade,
+	createSnapshotAsynchronously func(ctx context.Context, chainID string, toHeight int64) (snapshot.MetaData, error),
+	snapshotPublished func(meta snapshot.MetaData) (bool, error),
+) *ProtocolUpgradeHandler {
+	return &ProtocolUpgradeHandler{
+		log:                          log.Named("protocol-upgrade-handler"),
+		cfg:                          cfg,
+		protocolUpgradeService:       protocolUpgradeService,
+		createSnapshotAsynchronously: createSnapshotAsynchronously,
+		snapshotPublished:            snapshotPublished,
+	}
+}
+
+func (t *ProtocolUpgradeHandler) OnProtocolUpgradeEvent(ctx context.Context, chainID string,
+	lastCommittedBlockHeight int64,
+) (snapshot.MetaData, error) {
+	snapshotMetaData, err := t.createSnapshotAsynchronously(ctx, chainID, lastCommittedBlockHeight)
+	if err != nil && !errors.Is(err, snapshot.ErrSnapshotExists) {
+		return snapshot.MetaData{}, fmt.Errorf("failed to create snapshot:%w", err)
+	}
+
+	t.log.Infof("waiting for snapshot to be published, maximum wait time %s, snapshot %+v",
+		t.cfg.ProtocolUpgradeTimeout.Duration, snapshotMetaData)
+
+	ctxWithTimeout, cancelFn := context.WithTimeout(ctx, t.cfg.ProtocolUpgradeTimeout.Duration)
```

Comment:
I took the view that there is a meaningfully higher probability of issues occurring during protocol upgrade and therefore it would be prudent to have the last pre upgrade segment both created AND published to give us the most recovery options (for example you can restore from dehistory if that were needed in some unforeseen scenario)  But maybe being overzealous you think?

Reply:
I don't really mind whether it gets published or not, I just thought maybe not publishing it would make it a bit more simple. 

I think as long as its all synchronous it'll be fine because it'll help minimise the downtime during the upgrade. For example if a node operator sets their `AddSnapshotsInterval` to `1hr` for whatever reason, during upgrade the datanode could just be sitting their not shutting down for an hour waiting just waiting for that publish tick to happen.

Reply:
refactored the code to allow synchronous creation and publication